### PR TITLE
Ignore common label-values in ProxyResponseHeap sort function

### PR DESCRIPTION
**Pull request by wallee94**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Change `ProxyResponseHeap.Less` function to compare only the label-values that are not in both series. 
Some components are sorting the response before adding external labels, returning sets that might not be sorted after that. Because common label-values don't make a difference when comparing two label sets, ignoring these common external labels can help sort the response in the heap again.

I described the bug and how to replicate in https://github.com/thanos-io/thanos/issues/6257#issuecomment-1514160495

## Verification

I added a few tests for `ProxyResponseHeap` that I used to replicate the bug and test the changes.

### CODE REVIEW

File: pkg/store/proxy_heap_test.go [hunk 1]
```diff
@@ -22,6 +22,123 @@ func TestRmLabelsCornerCases(t *testing.T) {
...
+func (l *testingRespSet) Empty() bool {
+	return l.i >= len(l.bufferedResponse)
+}
+
+func TestProxyResponseHeapSort(t *testing.T) {
+	for _, tcase := range []struct {
+		input []respSet
+		exp   []*storepb.SeriesResponse
+	}{
+		// Different sorted response sets
```
Comment (fpetkovski) [inline pkg/store/proxy_heap_test.go:58]: Maybe we can convert this description to a `name` field and use it in `t.Run(`. This way the IDE can help out with running individual test cases.

**Review by fpetkovski — COMMENTED** [review]
Thanks for taking a stab at this!

I think a better way would be to exclude external labels from the comparison function since they are the ones that affect the sorting order. Because they are know in advance, we can construct a set of labels once per series request instead of once per comparison. The latter will likely be far more expensive.

**Comment by douglascamata** [comment]
Is this and #6296 achieving the same outcome, but through different means? 🤔

**Comment by fpetkovski** [comment]
I think we need both for this issue

File: pkg/store/proxy_heap_test.go (hunk 1 above)
Comment (wallee94) [reply pkg/store/proxy_heap_test.go]: Sounds good! Added

**Comment by wallee94** [comment]
Ah I didn't realize the first time that `st.LabelSets()` were the external labels in the store. I added a field to store them in advance.
Also, did I break the end-to-end tests? :thinking:

File: pkg/store/proxy_heap_test.go [hunk 2]
```diff
@@ -22,6 +22,132 @@ func TestRmLabelsCornerCases(t *testing.T) {
 	}), labels.Labels{})
 }
 
+type testingRespSet struct {
```
Comment (fpetkovski) [inline pkg/store/proxy_heap_test.go:25]: Can we reuse the mocks and stubs from the `testutil` package? You can see an example usage [here](https://github.com/thanos-io/thanos/blob/main/pkg/store/proxy_test.go#L79).

File: pkg/store/proxy_heap.go [hunk 3]
```diff
@@ -170,9 +170,13 @@ func (h *ProxyResponseHeap) Less(i, j int) bool {
...
 
 	if iResp.GetSeries() != nil && jResp.GetSeries() != nil {
+		// Response sets are sorted before adding external labels.
+		// This comparison excludes those labels to keep the same order.
+		iStoreLbls := (*h)[i].rs.StoreLabels()
+		jStoreLbls := (*h)[j].rs.StoreLabels()
 		iLbls := labelpb.ZLabelsToPromLabels(iResp.GetSeries().Labels)
 		jLbls := labelpb.ZLabelsToPromLabels(jResp.GetSeries().Labels)
-		return labels.Compare(iLbls, jLbls) < 0
+		return labels.Compare(rmLabels(iLbls.Copy(), iStoreLbls), rmLabels(jLbls.Copy(), jStoreLbls)) < 0
```
Comment (fpetkovski) [inline pkg/store/proxy_heap.go:179]: I think we need to compare the full label set when the comparison without external labels is 0.

Comment (wallee94) [reply pkg/store/proxy_heap.go]: You are right. One example I see is when the responses have repeated series, but their stores had different external labels. it has to compare the full set to keep the repeated series together.
I have added a test with this case.

File: pkg/store/proxy_heap_test.go (hunk 2 above)
Comment (wallee94) [reply pkg/store/proxy_heap_test.go]: I could not find an implementation of `respSet` in `testutil`, but I changed it to use one from proxy_head.go.

**Comment by fpetkovski** [comment]
Thanks for the changes 👍 I ran the proxy benchmark locally using `go test -bench BenchmarkProxySeries ./pkg/store/... -run none -count 5 -benchmem` and noticed that this implementation introduces a memory and CPU regression. I would suggest the following changes to reduce the impact a bit: https://github.com/wallee94/thanos/pull/1. Here are the results from the main branch, this branch and proposed changes:

```
thanos $ benchstat main.out branch.out branch-optimized.out                                           
name \ time/op                                                                          main.out     branch.out    branch-optimized.out
ProxySeries/1000000SeriesWith1Samples/4_client_with_1_samples,_250000_series_each-8      126ms ± 3%    250ms ±10%            159ms ± 2%
ProxySeries/1000000SeriesWith1Samples/single_client_with_1_samples,_1000000_series-8     177ms ± 1%    179ms ± 3%            168ms ± 0%
ProxySeries/100000SeriesWith100Samples/4_client_with_25_samples,_25000_series_each-8    13.5ms ± 4%   25.3ms ± 2%           16.5ms ± 2%
ProxySeries/100000SeriesWith100Samples/single_client_with_100_samples,_100000_series-8  16.9ms ± 0%   17.6ms ± 3%           17.9ms ± 1%
ProxySeries/1SeriesWith10000000Samples/4_client_with_2500000_samples,_1_series_each-8   16.4µs ± 2%   16.8µs ± 2%           16.4µs ± 1%
ProxySeries/1SeriesWith10000000Samples/single_client_with_10000000_samples,_1_series-8  7.53µs ± 1%   7.82µs ± 1%           7.74µs ± 2%

name \ alloc/op                                                                         main.out     branch.out    branch-optimized.out
ProxySeries/1000000SeriesWith1Samples/4_client_with_1_samples,_250000_series_each-8     83.9MB ± 0%  243.9MB ± 0%           83.9MB ± 0%
ProxySeries/1000000SeriesWith1Samples/single_client_with_1_samples,_1000000_series-8    83.4MB ± 0%   83.4MB ± 0%           83.4MB ± 0%
ProxySeries/100000SeriesWith100Samples/4_client_with_25_samples,_25000_series_each-8    7.64MB ± 0%  23.64MB ± 0%           7.64MB ± 0%
ProxySeries/100000SeriesWith100Samples/single_client_with_100_samples,_100000_series-8  8.21MB ± 0%   8.21MB ± 0%           8.21MB ± 0%
ProxySeries/1SeriesWith10000000Samples/4_client_with_2500000_samples,_1_series_each-8   7.85kB ± 0%   8.81kB ± 0%           8.23kB ± 0%
ProxySeries/1SeriesWith10000000Samples/single_client_with_10000000_samples,_1_series-8  3.52kB ± 0%   3.71kB ± 0%           3.77kB ± 0%

name \ allocs/op                                                                        main.out     branch.out    branch-optimized.out
ProxySeries/1000000SeriesWith1Samples/4_client_with_1_samples,_250000_series_each-8        302 ± 0%  2500311 ± 0%              308 ± 0%
ProxySeries/1000000SeriesWith1Samples/single_client_with_1_samples,_1000000_series-8       152 ± 0%      154 ± 0%              154 ± 0%
ProxySeries/100000SeriesWith100Samples/4_client_with_25_samples,_25000_series_each-8       251 ± 0%   250259 ± 0%              257 ± 0%
ProxySeries/100000SeriesWith100Samples/single_client_with_100_samples,_100000_series-8     132 ± 0%      134 ± 0%              134 ± 0%
ProxySeries/1SeriesWith10000000Samples/4_client_with_2500000_samples,_1_series_each-8      130 ± 0%      146 ± 0%              136 ± 0%
ProxySeries/1SeriesWith10000000Samples/single_client_with_10000000_samples,_1_series-8    78.0 ± 0%     80.0 ± 0%             80.0 ± 0%
```

If we want to completely the impact of ignoring external labels, we could probably add them in a separate field when sending data from stores, but for now we should be okay with a small regression.

**Comment by wallee94** [comment]
Oh wow, I didn't expect so much memory regression. The new changes look great to me, I've merged them and the latest changes in main. Thanks again @fpetkovski !

**Review by fpetkovski — APPROVED** [review]

**Comment by fpetkovski** [comment]
@wallee94 have you checked whether this change solves your problem?

**Comment by wallee94** [comment]
@fpetkovski Yes! I deployed the branch today to our dev cluster. We were seeing duplicated series from the ruler and sidecars, and this change fixed the issue

**Pull request merged**